### PR TITLE
[pallas] Fix the interpreter for block_shape not dividing the overall shape

### PR DESCRIPTION
**Pull request by gnecula**
Before this change, the interpreter was failing with an MLIR verification error because the body of the while loop returned a padded output array.

This change allows us to expand the documentation of block specs with the case for when block_shape does not divide the overall shape.

### CODE REVIEW

File: jax/_src/pallas/pallas_call.py [hunk 1]
```diff
@@ -164,142 +164,141 @@ def _get_next_indices(grid, indices):
...
+    grid_mapping: GridMapping,
+    compiler_params: Any):
+  del compiler_params, name, in_shapes
+  # If we're in interpreter mode, we *scan* over the grid and eval the
+  # discharged jaxpr. This should reproduce exactly what compiling to Triton
+  # will do.
   dynamic_grid_args, args = split_list(  # type: ignore
       args, [grid_mapping.num_dynamic_grid_bounds]
   )
-  if interpret:
```
Comment (gnecula) [inline jax/_src/pallas/pallas_call.py:176]: This code block was just unindented

File: jax/_src/pallas/pallas_call.py [hunk 2]
```diff
@@ -164,142 +164,141 @@ def _get_next_indices(grid, indices):
...
+  for o, expected_o_shape, bm in zip(out, out_shapes, out_block_mappings):
+    if bm is not None and isinstance(bm.indexing_mode, pallas_core.Unblocked):
+      padding = bm.indexing_mode.padding
+      if padding is not None and any(p != (0, 0) for p in padding):
+        if input_output_aliases:
+          raise NotImplementedError("Padding with aliasing not supported.")
+        pad_low, pad_high = zip(*padding)
+        limit_indices = [s - p for s, p in zip(o.shape, pad_high)]
+        o = lax.slice(o, pad_low, limit_indices)
+    if o.shape != expected_o_shape.shape:
```
Comment (gnecula) [inline jax/_src/pallas/pallas_call.py:297]: This is the only new part

File: jax/_src/pallas/pallas_call.py [hunk 3]
```diff
@@ -164,142 +164,141 @@ def _get_next_indices(grid, indices):
...
+      None if bm is None else tuple(b is pallas_core.mapped for b in bm)
+      for bm in block_shapes_without_mapped_dims
+  ]
+  block_shapes = [
+      None if (bm is None or iid is None)
+      else tuple(1 if i else b for i, b in zip(iid, bm))
+      for iid, bm in zip(is_indexing_dim, block_shapes_without_mapped_dims)
+  ]
 
-    # Pad values to evenly divide into block dimensions.
```
Comment (gnecula) [inline jax/_src/pallas/pallas_call.py:230]: Also just unindented

File: jax/_src/pallas/pallas_call.py [hunk 4]
```diff
@@ -164,142 +164,141 @@ def _get_next_indices(grid, indices):
...
+    i, *_ = carry
+    return i < num_iterations
+  def body(carry):
+    i, loop_idx, *carry = carry
+    local_grid_env = tuple(
+        pallas_core.GridAxis(idx, b)
+        for dim, (idx, b) in enumerate(zip(loop_idx, grid))
+        if dim not in grid_mapping.mapped_dims
     )
-    _, out, _ = split_list(carry, [len(args), len(out)])
```
Comment (gnecula) [inline jax/_src/pallas/pallas_call.py:281]: Also unindented

File: jax/_src/pallas/pallas_call.py [hunk 5]
```diff
@@ -164,142 +163,141 @@ def _get_next_indices(grid, indices):
...
+    grid_mapping: GridMapping,
+    compiler_params: Any):
+  del compiler_params, name, in_shapes
+  # If we're in interpreter mode, we *scan* over the grid and eval the
+  # discharged jaxpr. This should reproduce exactly what compiling to Triton
+  # will do.
   dynamic_grid_args, args = split_list(  # type: ignore
       args, [grid_mapping.num_dynamic_grid_bounds]
   )
-  if interpret:
```
Comment (gnecula) [inline jax/_src/pallas/pallas_call.py:176]: Almost all of the below is now unindented.
This is easier to review with the internal tooling: https://critique.corp.google.com/cl/649422123

File: tests/pallas/pallas_test.py [hunk 6]
```diff
@@ -458,6 +458,29 @@ def softmax_kernel(x_ref, y_ref):
 
     np.testing.assert_allclose(softmax_kernel(x), jax.nn.softmax(x), atol=1e-7)
 
+  def test_block_spec_with_padding(self):
+    self.INTERPRET = True
```
Comment (superbobry) [inline tests/pallas/pallas_test.py:462]: Maybe add it to the `PallasCallInterpreterTest` below?

File: jax/_src/pallas/pallas_call.py [hunk 7]
```diff
@@ -164,142 +163,141 @@ def _get_next_indices(grid, indices):
...
+    *args,
+    jaxpr: jax_core.Jaxpr,
+    name: str, in_shapes, out_shapes,
+    debug: bool,
+    input_output_aliases: tuple[tuple[int, int], ...],
+    grid_mapping: GridMapping,
+    compiler_params: Any):
+  del compiler_params, name, in_shapes
+  # If we're in interpreter mode, we *scan* over the grid and eval the
+  # discharged jaxpr. This should reproduce exactly what compiling to Triton
```
Comment (superbobry) [inline jax/_src/pallas/pallas_call.py:179]: Nit: consider rephrasing because we are not just targeting Triton now.

File: jax/_src/pallas/pallas_call.py [hunk 8]
```diff
@@ -164,142 +163,141 @@ def _get_next_indices(grid, indices):
     next_indices.append(jnp.where(carry, 0, i))
   return tuple(reversed(next_indices))
 
-def _pallas_call_impl(*args, jaxpr, name, out_shapes,
-                      interpret, debug: bool,
-                      in_shapes,
-                      input_output_aliases: tuple[tuple[int, int], ...],
-                      grid_mapping: GridMapping,
-                      compiler_params: Any):
+def _pallas_call_impl(*args, **kwargs):
```
Comment (superbobry) [inline jax/_src/pallas/pallas_call.py:166]: Can we avoid calling `def_impl` entirely and get an error from the Primitive itself (since `impl` will be `None`)?

**Review by superbobry — APPROVED** [review]

File: tests/pallas/pallas_test.py (hunk 6 above)
Comment (gnecula) [reply tests/pallas/pallas_test.py:462]: This was left over from my debugging. Removed.

File: jax/_src/pallas/pallas_call.py (hunk 8 above)
Comment (gnecula) [reply jax/_src/pallas/pallas_call.py:166]: I prefer it this way: it adds minimal code but allows me to add a comment next to "assert False" of why we do not need an impl rule.

File: docs/pallas/grid_blockspec.md [hunk 9]
```diff
@@ -104,11 +104,11 @@ the overall array). Each block index is then multiplied by the
...
-```{note}
-This documentation applies to the case when the block shape divides
-the array shape.
-The documentation for the other cases is pending.
-```
+If the block shape does not divide evenly the overall shape then the
+last iteration on each axis will still receive references to blocks
+of `block_shape` but the elements that are out-of-bounds are padded
+on input and discarded on output. Note that at least one of the
+elements in each block must be within bounds.
```
Comment (apaszke) [inline docs/pallas/grid_blockspec.md:111]: It would be good to mention that the padding value is unspecified

Comment (gnecula) [reply docs/pallas/grid_blockspec.md:111]: Thanks, will fix in a future PR, this is being submitted
